### PR TITLE
Adding optional `Tracer` to `reverseproxy.Proxy`.

### PR DESCRIPTION
**Pull request by dhermes**
## PR Summary

 - **Type:** Features
 - **Intended Change Level:** minor

**NOTE**: This has #164 as diffbase (though since it's in my fork, I can't make it the diffbase of this PR). As of 3:15pm on January 15, 2020 [this][1] view will give the 4 commits relevant to this PR.

[1]: https://github.com/blend/go-sdk/pull/165/files/95f2209b3917ed6ca2dae6b5196e8a2e5230926d..02c1875cfed72deb264691d4a6c3f2dae22f8ae7

### CODE REVIEW

File: reverseproxy/proxy.go [hunk 1]
```diff
@@ -45,6 +47,16 @@ func (p *Proxy) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
...
 	}()
 
+	var err error
+	if p.Tracer != nil {
+		var tf web.HTTPTraceFinisher
+		tf, req = p.Tracer.Start(req)
+		// NOTE: It is crucial to use `err` in a closure from current scope,
+		//       **not** as an argument. This way updates to `err` will be
+		//       reflected on exit.
+		defer func() { tf.Finish(err) }()
```
Comment (wcharczuk) [inline reverseproxy/proxy.go]: this will not capture the cause of panics, is that intentional?

Comment (dhermes) [reply reverseproxy/proxy.go]: Hadn't considered panics. I can put this into the current panic handler if you think it makes sense, e.g.

```diff
diff --git a/reverseproxy/proxy.go b/reverseproxy/proxy.go
index 4712e8a..d9e261c 100644
--- a/reverseproxy/proxy.go
+++ b/reverseproxy/proxy.go
@@ -5,6 +5,7 @@ import (
 	"net/http"
 	"os"
 
+	"github.com/blend/go-sdk/ex"
 	"github.com/blend/go-sdk/logger"
 	"github.com/blend/go-sdk/webutil"
 )
@@ -36,24 +37,36 @@ type Proxy struct {
 
 // ServeHTTP is the http entrypoint.
 func (p *Proxy) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	var err error
+	var tf webutil.HTTPTraceFinisher
 	defer func() {
-		if r := recover(); r != nil {
-			if p.Log != nil {
-				p.Log.Fatalf("%v", r)
-			} else {
-				fmt.Fprintf(os.Stderr, "%v\n", r)
-			}
+		r := recover()
+
+		// Wrap the error with the reason for the panic.
+		if rErr, ok := r.(error); ok {
+			err = ex.Nest(err, rErr)
+		} else {
+			err = ex.Nest(err, ex.New(r))
+		}
+
+		// Finish the span, if open
+		if tf != nil {
+			tf.Finish(err)
+		}
+
+		// Log or print the cause of the panic.
+		if r == nil {
+			return
+		}
+		if p.Log != nil {
+			p.Log.Fatalf("%v", r)
+		} else {
+			fmt.Fprintf(os.Stderr, "%v\n", r)
 		}
 	}()
 
-	var err error
 	if p.Tracer != nil {
-		var tf webutil.HTTPTraceFinisher
 		tf, req = p.Tracer.Start(req)
-		// NOTE: It is crucial to use `err` in a closure from current scope,
-		//       **not** as an argument. This way updates to `err` will be
-		//       reflected on exit.
-		defer func() { tf.Finish(err) }()
 	}
 
 	// set the default resolver if unset.
```

Comment (wcharczuk) [reply reverseproxy/proxy.go]: i would have the tracer capture panics, yes

**Comment by dhermes** [comment]
@wcharczuk PTAL

**Pull request merged**